### ISSUE TITLE
FR: [calibre]: get updated metadata without pushing books again

### ISSUE THREAD

**Issue by pazos** [koreader/koreader#12922]
Reported in https://www.reddit.com/r/Calibre/comments/1hfymg6/metadata_is_never_updated_on_device_smartdevice/

I took a look. `SEND_BOOK_METADATA` is not implemented.

Currently the only way to push metadata to device after edit is resending the books again.
I didn't investigate, but looks like we need to handle the callback more or less like `SEND_BOOK`

**Comment by emrsmsrli** [comment]
hey, any update on this? I'd love to contribute but I have no idea where to start actually

**Comment by pazos** [comment]
1. Add the `OPCODE` to https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/wireless.lua#L34
2. Update https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/wireless.lua#L391 to react to the OPCODE calling a new function.
3. Most opcodes receive an argument in the form of a table (all except NOOP).
4. The logic in the new function will be, most probably, like https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/wireless.lua#L621, except we don't need to push files, so much easier to implement.

**Comment by emrsmsrli** [comment]
this is certainly harder than I thought. metadata is stored in multiple places, and I don't know if it's intentional or not. I managed to make it work for calibre metadata searching and regular file browser, but coverbrowser plugin is really hard to tackle because it stores custom data along with book data. also this solution is very brittle because it doesn't recover from "reset cached metadata" actions, or if the user deletes the sidecard folder.

maybe the whole logic for storing metadata needs to be refactored, what do you think @pazos? 

currently what I could found is this:
- **metadata.lua** file
  - under **doc_props**
  - under **stats** (but it's partial and if _stats_ plugin is enabled?)
- **custom_metadata.lua** file under **doc_props**
- in a database (if _coverbrowser_ plugin is enabled)
- in **.metadata.calibre** file (if the book is coming from calibre, via the _calibre_ plugin)
- in the book itself

ideally the fix for this issue should be patching the book itself, and resetting all the metadata locations. but I'm not sure if patching is possible in koreader side

**Comment by hius07** [comment]
(I know nothing about the calibre plugin)
Is updated metadata written into the book file?
If yes, the new book file is needed.
If not, the updated metadata should be saved as custom metadata.
We do this in
https://github.com/koreader/koreader/blob/c3352ee171aa8ea695f64518a38c03b759e7cb24/frontend/apps/filemanager/filemanagerbookinfo.lua#L468

**Comment by emrsmsrli** [comment]
@hius07 yes, calibre saves the new metadata in the file itself. and then when koreader connects to it, the metadata in calibre is compared to the one in koreader _.metadata.calibre_ file. if they differ, calibre sends the new metadata in json format. 

in my opinion at this stage koreader should either patch the book file on device, or ask the whole file again from calibre (I don't know if this would have any side effects)

in this context, custom metadata means the user put something in though, right? I'm not sure if this new metadata from calibre should go in there

**Comment by pazos** [comment]
calibre metadata is stored in json files. When a user updates a metadata field on a book (present on device's library) via calibre and clicks the "refresh device metadata" we're expected to react to the `SEND_BOOK_METADATA` by updating (calibre-metadata) book fields that were changed and store that as a json file.

These calibre json files are only used to "search calibre metadata" and by the wireless client itself.

The rest of the program relies on real metadata buried into files. That's what we get if the user pushes the new version of the book again (but that's another intent, covered by `SEND_BOOK`)

I wouldn't mind improvements in the plugin later (ie: auto-collections based on metadata fields) but as far as the wireless client is regarded I think we should react to the callback accordly and update our own records but never touch on-device files.

**Comment by emrsmsrli** [comment]
json file part is easy. but as I said modifying metadata.lua files in the sidecar folder is brittle because the will be overwritten by the book metadata at some point. based on this, your recommendation is to override whatever's in the custom metadata file then?

**Comment by pazos** [comment]
> json file part is easy. but as I said modifying metadata.lua files in the sidecar folder is brittle because the will be overwritten by the book metadata at some point. based on this, your recommendation is to override whatever's in the custom metadata file then?

No modification of metadata.lua. No modification at all other than calibre json files.

At least for this feature request.

**Comment by pazos** [comment]
In case I wasn't clear enough.

The source of truth for calibre metadata are calibre json files (updated by calibre on usb transfers, updated by us on "smart device app")

The source of truth for the rest of the program are the files themselves. 

Trying to fixing mismatches between calibre metadata and real metadata as present in files is outside the scope of this FR. It doesn't make much sense when there's a way to do that from calibre using another, different callback.

**Comment by emrsmsrli** [comment]
I got your point, the thing is, the reddit post mentions sending the book again will cause annotations to disappear (I haven't tested this myself). And I do believe rather than sending the whole file, just patching the existing one on device would be the most efficient solution, long-term.

**Comment by pazos** [comment]
reddit is not a source of trust :)

**Edit**: updated OP with a quote of the issue we're trying to fix here.

> sending the book again will cause annotations to disappear

It has nothing to do with metadata. If you send a different version of the book you might screw annotations, bookmarks and things that rely on the DOM. That's not an issue. That's an user shooting itself on the foot. 


> And I do believe rather than sending the whole file, just patching the existing one on device would be the most efficient solution, long-term.

That's not the intent of `SEND_BOOK_METADATA`.
Please don't try to outsmart the protocol here :)

Other contributions might be welcome. Feel free to research past tickets about the plugin and open as many feature requests as you want that were not covered yet.
